### ISSUE TITLE
HTTP Parse Error when POST request has `Transfer-Encoding: chunked`

### ISSUE THREAD

**Issue by ademars94** [puma/puma#1492]
### Steps to reproduce

1) Clone this repo: https://github.com/ademars94/puma-test

2) Follow the installation instructions.

3) Follow the test instructions and see error in console. OR you can use a rest client of some sort (e.g. Postman or similar) to send a `POST` request to `localhost:9292/foo` with the HTTP header `Transfer-Encoding: chunked`

### Expected behavior

There shouldn't be an HTTP Parse Error, and the request body should be complete in the controller layer.

### Actual behavior

There is an HTTP Parse Error.  Sometimes the request body, by the time it gets to the controller layer, will be missing some chunks, resulting in invalid data.

### System configuration

**Ruby version**: 2.4
**Rails version**: N/A, but this error occurs in Rails 5.1.x as well
**Sinatra version**: 2.0
**Puma version**: 3.9

**Comment by evgeniy-kunitsa** [comment]
I have the same problem. I wrote simple rack app to save image which was received via "chunked" request. But some data was lost.

`require 'rack'

class Proxy
  def call(env)
    f = File.open("#{Time.now}.jpg", 'wb')
    f.write(env['rack.input'].read)

    puts f.size
    f.close

    success
  end
  
  def success
    [200, {}, ['Success!']]
  end
end

run Proxy.new`

New file size is 4623975, but original file size is 4623987.
Command to send file with chunked encoding:
`curl -v -X POST --header "Transfer-Encoding: chunked" --data-binary @original.jpg "http://localhost:9292/"`

## System configuration
**Ruby version:** 2.4.2
**Rack version:** 2.0.3
**Puma version:** 3.11

**Comment by sheldonh** [comment]
Confirmed in 3.11.2.

I think the problem is that the `Tempfile` is rewound but not flushed in `Puma::Server#read_body`.

**Comment by sheldonh** [comment]
This is hard to test. I The test below passed 284 times before failing!

The test:

```
  def test_chunked_large_request
    bytes = 0
    reads = 0
    @server.app = proc { |env|
      while buf = env['rack.input'].read(4096)
        reads += 1
        bytes += buf.size
      end
      [200, {}, [""]]
    }

    @server.add_tcp_listener @host, @port
    @server.run

    sock = TCPSocket.new @host, @server.connected_port
    sock << "GET / HTTP/1.1\r\nConnection: close\r\nTransfer-Encoding: chunked\r\n\r\n"
    64.times { sock << "4000\r\n"; sock << "#{'x' * 0x4000}\r\n" } # Net::HTTP
    sock << "0\r\n\r\n"

    data = sock.read

    assert_equal "HTTP/1.1 200 OK\r\nConnection: close\r\nContent-Length: 0\r\n\r\n", data
    assert_equal 64 * 0x4000, bytes
    assert_equal 256, reads
  end
```

The failure:

```
$ i=0; while ruby test/test_puma_server.rb --name test_chunked_large_request; do i=$(($i + 1)); sleep 0.001; done; echo $i tests completed successfully
...
  1) Failure:
TestPumaServer#test_chunked_large_request [test/test_puma_server.rb:581]:
Expected: 1048576
  Actual: 1052672

1 runs, 2 assertions, 1 failures, 0 errors, 0 skips
283 tests completed successfully
```

When I add a flush before rewind to the `Tempfile` in `Puma::Server#read_body`, the test still runs successfully most of the time, but then it hangs instead of failing.

**Comment by ademars94** [comment]
Update: This is still an issue on v3.11.x, and the test from @sheldonh does seem to reproduce the bug (although it only fails sometimes).

**Comment by nateberkopec** [comment]
Fixed in #1812/4.1.

**Comment by rngtng** [comment]
thx
